doc/specifications/proposed/replicator/architecture.rst

```diff
@@ -0,0 +1,142 @@
+==========
+Replicator
+==========
+
+This document describes a common synchronization mechanism between two systems,
+which can cope with write failures, network downtimes and other types of
+partitions. The approach only works with a single write master (Single Source
+of Truth) and any number of read slaves.
+
+In our case we want to use this approach to synchronize the content database
+with a search index.
+
+General Workings
+================
+
+The general scheme for synchronizations is fairly simple:
+
+.. image:: synchronize.png
+
+It generally can be described as a three step process:
+
+1) Fetch last revision from target
+
+   The target is asked for the last revision the target knows. This is the
+   revision of the last successful write operation. For the first request the
+   target may return something like ``null``.
+
+2) Fetch all changes since that last revision from source
```

Comment:
NB: this means that master should store more or less indefinitely all data that it normally discards. This might pose a storage problem.

Eg. version 1-100 of object X, which is then deleted at revision 101 have to be kept, most likely in a separate storage space, until all slaves have duplicated them.

In the legacy content engine a lot of operations are permitted which allow the editor to destroy historical content, eg. remove old revisions, remove languages (which alters old revisions), remove contents.